### DISCUSSION TITLE
How does server-side validation work?

### DISCUSSION THREAD

**Discussion by tanmaymishu** [pouchdb/pouchdb#8518] [Q&A]
Ok, I am only 1-day old in the PouchDB/Offline-first realm and this might sound like a dumb question. I so far understand what its intended use cases are and how it generally works. But I have a question. How does the validation fit into that picture? If the PouchDB source is loaded in the browser and some malicious code is executed, let's say to override a user's role from 'editor' to 'admin' wouldn't that be synced with the server-side replica as well? In the online-first world, typically the server-side application validates the incoming data, how does that work in the PouchDB world?

**Comment by Terreii** [comment]
Not a dumb question at all.

The server-side is usually done with [CouchDB](https://couchdb.apache.org/). It is then responsible for validation and authorization.

**Authorization**
CouchDB has a special db, the [`_users` db](https://docs.couchdb.org/en/stable/intro/security.html), in it you store your user data. Then every db in CouchDB has a [`/{db}/_security`](https://docs.couchdb.org/en/stable/api/database/security.html) API where admins can give access to users (by name or by role).

**Validation**
Then in every db there are [_design docs](https://docs.couchdb.org/en/stable/ddocs/index.html). They store secondary indexes and [_validations_](https://docs.couchdb.org/en/stable/ddocs/ddocs.html#validate-document-update-functions). A document _must_ pass all _Validate Document Update Functions_ in a db, to be accepted.

You could also add a proxy between the client and CouchDB using [express-pouchdb](https://www.npmjs.com/package/express-pouchdb) or any other proxy package and do your own validation.

**Security**
There are 2 types of admins in CouchDB:
- Admin users
- DB admins

**Admin users** are created by [config](https://docs.couchdb.org/en/stable/intro/security.html#creating-a-new-admin-user). They are _not_ stored in the `_users` db. CouchDB doesn't even start without one Admin user. They have absolute power. They can change configurations, read and write all user data, and have db-admin rights to all dbs.
**DB admins** are users that have the right to write ddocs (design docs). Their right is given by db, in the [`/{db}/_security` API](https://docs.couchdb.org/en/stable/api/database/security.html).

PouchDB by itself doesn't have the notion of authorization or validation. For client-side reuse of the validate-update-functions you could use the [pouchdb-validation](https://www.npmjs.com/package/pouchdb-validation) plugin.
Also a user could add a ddoc. But if you didn't give them admin rights to a db, then it will never be synced to other users, nor accepted by CouchDB.
